training/ikon20/part2-working-with-masks.ipynb

```diff
@@ -0,0 +1,256 @@
+   "outputs": [],
+   "source": [
+    "data = sc.io.open_hdf5(filename='loki-at-larmor.hdf5')\n",
+    "counts = sc.sum(data, 'tof') # used later\n",
+    "data"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Masks are variables with `dtype=bool`, stored in the `masks` dict of a data array.\n",
+    "The result of comparison between variables can thus be used as masks:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "data.coords['spectrum'] < sc.scalar(100)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Exercise 1: Masking a prompt pulse\n",
+    "\n",
+    "1. Create a promt-pulse mask for the region between $17500~\\mathrm{\\mu s}$ and $19000~\\mathrm{\\mu s}$.\n",
+    "   Notes:\n",
+    "   - Use comparison operators such as `==`, `<=` or `>`.\n",
+    "   - Combine multiple conditions into one using `&` (\"and\"), `|` (\"or\"), or `^` (\"exclusive or\").\n",
+    "   - Masks are stored in a data array by storing them in the `masks` dictionary, e.g., `data.masks['promt-pulse'] = ...`.\n",
+    "   - If something goes wrong, masks can be removed with Python's `del`, e.g., `del data.masks['wrong']`.\n",
+    "   - If you run into an error regarding a length mismatch when inserting the coordinate, remember that `'tof'` is a bin-edge coordinate, i.e., it is by 1 longer than the number of bins.\n",
+    "     Use, e.g., only the left bin edges, i.e., all but the last, to create the masks.\n",
+    "2. Use the HTML view and plot the data after masking to explore the effect.\n",
+    "3. Pass a `dict` containing `counts` (computed above as `counts = sc.sum(data, 'tof')`) and the equivalent counts computed *after* masking to `sc.plot.plot`.\n",
+    "   Use this to verify that the promt-pulse mask results in removal of counts."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "tof = data.coords['tof']\n",
+    "data.masks['promt-pulse'] = (tof['tof',:-1] > 17500.0 * sc.units.us) & (tof['tof',1:] < 19000.0 * sc.units.us)\n",
+    "sc.plot.plot({'before':counts, 'after':sc.sum(data, 'tof')})"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Exercise 2: Masking spatially\n",
+    "\n",
+    "By masking an `x` range, mask the end of the tubes.\n",
+    "- Define `x = sc.geometry.x(data.coords['position'])` to extract only the x-component of the position coordinate.\n",
+    "- Create the masks.\n",
+    "- Use the instrument view (`sc.neutron.instrumentview(data)`) to inspect the result."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "x = sc.geometry.x(data.coords['position'])\n",
+    "data.masks['tube-ends'] = x < -0.2 * sc.units.m\n",
+    "sc.neutron.instrument_view(data)"
```

Comment:
Do you need to add the `pixel_size` again?

Reply:
Will decide once we know if the server is running lab